# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1744485904 -7200
# Parent  0779eb99e08d5062b476de9c5676e740154a9043
Bug 695142 - Context menu to open link in browser in compose window.
* * *
Pay attention to pref mail.external_protocol_requires_permission when opening non-http(s) links in compose window.

diff --git a/mail/components/compose/content/MsgComposeCommands.js b/mail/components/compose/content/MsgComposeCommands.js
--- a/mail/components/compose/content/MsgComposeCommands.js
+++ b/mail/components/compose/content/MsgComposeCommands.js
@@ -45,8 +45,12 @@ var { AppConstants } = ChromeUtils.impor
 var { ExtensionParent } = ChromeUtils.importESModule(
   "resource://gre/modules/ExtensionParent.sys.mjs"
 );
 
+var { PlacesUtils } = ChromeUtils.importESModule(
+  "resource://gre/modules/PlacesUtils.sys.mjs"
+);
+
 ChromeUtils.defineESModuleGetters(this, {
   AttachmentInfo: "resource:///modules/AttachmentInfo.sys.mjs",
   BondOpenPGP: "chrome://openpgp/content/BondOpenPGP.sys.mjs",
   EnigmailKeyRing: "chrome://openpgp/content/modules/keyRing.sys.mjs",
@@ -1933,8 +1937,52 @@ function showMessageComposeSecurityStatu
     );
   }
 }
 
+var lastUrlHovered;
+var lastUriHovered;
+function openLinkInBrowser() {
+  if (!lastUrlHovered || !lastUriHovered) {
+    return;
+  }
+  PlacesUtils.history
+    .insert({
+      url: lastUrlHovered,
+      visits: [
+        {
+          date: new Date(),
+        },
+      ],
+    })
+    .catch(console.error);
+
+  const scheme = lastUriHovered.scheme;
+  const isHttp = ["http", "https"].includes(scheme);
+  const promptForPermission =
+    !isHttp &&
+    Services.prefs.getBoolPref("mail.external_protocol_requires_permission");
+
+  let principal = undefined;
+  if (!promptForPermission) {
+    // Copied from getContentPrincipalWithProtocolPermission() in LinkHelper.sys.mjs.
+    principal = Services.scriptSecurityManager.createContentPrincipal(
+      Services.io.newURI(
+        "chrome://messenger/content/messengercompose/messengercompose.xhtml"
+      ),
+      {}
+    );
+    Services.perms.addFromPrincipal(
+      principal,
+      `open-protocol-handler^${scheme}`,
+      Services.perms.ALLOW_ACTION
+    );
+  }
+
+  Cc["@mozilla.org/uriloader/external-protocol-service;1"]
+    .getService(Ci.nsIExternalProtocolService)
+    .loadURI(lastUriHovered, principal);
+}
+
 function msgComposeContextOnShowing(event) {
   if (event.target.id != "msgComposeContext") {
     return;
   }
@@ -2035,8 +2083,10 @@ function msgComposeContextOnShowing(even
 
   let onLink = false;
   let linkText = undefined;
   let linkUrl = undefined;
+  lastUrlHovered = null;
+  lastUriHovered = null;
 
   const link = target.closest("a");
   if (link) {
     onLink = true;
@@ -2046,9 +2096,14 @@ function msgComposeContextOnShowing(even
       link.getAttribute("a") ||
       link.href ||
       "";
     linkUrl = link.href;
-  }
+    try {
+      lastUriHovered = Services.io.newURI(linkUrl);
+      lastUrlHovered = linkUrl;
+    } catch (ex) {}
+  }
+  document.getElementById("openLinkInBrowser").hidden = !lastUriHovered;
 
   const subject = {
     menu: event.target,
     tab: window,
diff --git a/mail/components/compose/content/messengercompose.xhtml b/mail/components/compose/content/messengercompose.xhtml
--- a/mail/components/compose/content/messengercompose.xhtml
+++ b/mail/components/compose/content/messengercompose.xhtml
@@ -651,8 +651,9 @@
   <menuitem command="cmd_pasteNoFormatting"/>
   <menuitem label="&pasteQuote.label;" accesskey="&pasteQuote.accesskey;" command="cmd_pasteQuote"/>
   <menuitem data-l10n-id="text-action-delete" command="cmd_delete"/>
   <menuseparator/>
+  <menuitem id="openLinkInBrowser" label="&openLinkInBrowser.label;" oncommand="openLinkInBrowser();"/>
   <menuitem data-l10n-id="text-action-select-all" command="cmd_selectAll"/>
 
   <!-- Spellchecking general menu items (enable, add dictionaries...) -->
   <menuseparator id="spellCheckSeparator"/>
